### PR TITLE
Refactor CanvasKit backend in preparation for platform views

### PR DESCRIPTION
**Pull request by hterkelsen**
In this refactor I removed the distinction between "recording" SkCanvas and "normal" SkCanvas because in my platform view change I will be creating new SkCanvases for the drawing commands that need to go on top of the platform view.

I also refactored how we initialize SkSurfaces in preparation for allowing the platform view embedder to create new surfaces and overlay them with the root surface.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/compositor/util.dart [hunk 1]
```diff
@@ -56,6 +56,33 @@ js.JsArray<double> makeSkPoint(ui.Offset point) {
...
+
+js.JsObject makeSkPointMode(ui.PointMode pointMode) {
+  switch (pointMode) {
+    case ui.PointMode.points:
+      return canvasKit['PointMode']['Points'];
+    case ui.PointMode.lines:
+      return canvasKit['PointMode']['Lines'];
+    case ui.PointMode.polygon:
+      return canvasKit['PointMode']['Polygon'];
+  }
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/util.dart]: Throw if this escapes the switch block?

File: lib/web_ui/lib/src/engine/compositor/picture_recorder.dart [hunk 2]
```diff
@@ -5,30 +5,28 @@
...
-  ui.Rect cullRect;
-
+  ui.Rect _cullRect;
   js.JsObject _recorder;
 
-  @override
-  RecordingCanvas beginRecording(ui.Rect bounds) {
-    cullRect = bounds;
+  SkCanvas beginRecording(ui.Rect bounds) {
+    _cullRect = bounds;
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/picture_recorder.dart]: It's an estimate in Skia also. It's also used in the `preroll` for `PictureLayer` to estimate paint bounds.

File: lib/web_ui/lib/src/engine/compositor/canvas_kit_canvas.dart [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+
+part of engine;
+
+/// An implementation of [ui.Canvas] that is backed by a CanvasKit canvas.
+class CanvasKitCanvas implements ui.Canvas {
+  final SkCanvas _canvas;
+
+  factory CanvasKitCanvas(ui.PictureRecorder recorder, [ui.Rect cullRect]) {
+    assert(recorder != null);
+    if (recorder.isRecording)
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/canvas_kit_canvas.dart]: Agreed about preferring to use `{ }` even when not strictly necessary. This class is just a copy of the `Canvas` class from the mobile `dart:ui` but replacing every `native` call to a call to the `SkCanvas`

File: lib/web_ui/lib/src/engine/compositor/text.dart [hunk 4]
```diff
@@ -17,6 +17,9 @@ const Map<String, String> _fontFamilyOverrides = <String, String>{
   'AbrilFatface': 'Abril Fatface',
   'packages/cupertino_icons/CupertinoIcons': 'CupertinoIcons',
   '.SF Pro Text': 'Google Sans',
+  '.SF Pro Display': 'Google Sans',
+  '.SF UI Text': 'Google Sans',
+  '.SF UI Display': 'Google Sans',
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/text.dart]: I think the best way forward, which you mentioned the other day, is to try to get the family name declared in the actual file and map it to the name declared in the `FontManifest.json`

File: lib/web_ui/lib/src/engine/compositor/canvas_kit_canvas.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/canvas_kit_canvas.dart]: Done.

File: lib/web_ui/lib/src/engine/compositor/canvas_kit_canvas.dart [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+    assert(transforms != null);
+    assert(rects != null);
+    assert(colors != null);
+    assert(blendMode != null);
+    assert(paint != null);
+
+    final int rectCount = rects.length;
+    if (transforms.length != rectCount)
+      throw ArgumentError('"transforms" and "rects" lengths must match.');
+    if (colors.isNotEmpty && colors.length != rectCount)
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/canvas_kit_canvas.dart]: Done.

File: lib/web_ui/lib/src/engine/compositor/util.dart (hunk 1 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/util.dart]: Done.

**Pull request merged**